File: SparseArrays.py
#!/bin/python3

import math
import os
import random
import re
import sys

#
# Complete the 'matchingStrings' function below.
#
# The function is expected to return an INTEGER_ARRAY.
# The function accepts following parameters:
#  1. STRING_ARRAY strings
#  2. STRING_ARRAY queries
#

def matchingStrings(strings, queries):
    # Create a dictionary to count occurrences of each string
    string_count = {}

    # Count the occurrences of each string in the input list
    for string in strings:
        if string in string_count:
            string_count[string] += 1
        else:
            string_count[string] = 1

    # For each query, return the count from the dictionary (or 0 if not present)
    result = []
    for query in queries:
        result.append(string_count.get(query, 0))

    return result

if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    strings_count = int(input().strip())

    strings = []

    for _ in range(strings_count):
        strings_item = input()
        strings.append(strings_item)

    queries_count = int(input().strip())

    queries = []

    for _ in range(queries_count):
        queries_item = input()
        queries.append(queries_item)

    res = matchingStrings(strings, queries)

    fptr.write('\n'.join(map(str, res)))
    fptr.write('\n')

    fptr.close()

File: SparseArrays.java
import java.io.*;
import java.math.*;
import java.security.*;
import java.text.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.regex.*;

class Result
{
    public static List<Integer> matchingStrings(List<String> strings, List<String> queries)
    {
        HashMap<String, Integer> stringCount = new HashMap<String, Integer>();
        
        for (String str : strings)
        {
            if (stringCount.containsKey(str))
            {
                stringCount.put(str, stringCount.get(str) + 1);
            }
            else
            {
                stringCount.put(str, 1);
            }
        }
        
        List<Integer> result = new ArrayList<Integer>();
        
        for (String qry : queries)
        {
            if (stringCount.containsKey(qry))
            {
                result.add(stringCount.get(qry));
            }
            else
            {
                result.add(0);
            }
        }
        
        return result;
    }
}

public class Solution {
    public static void main(String[] args) throws IOException {
        BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(System.getenv("OUTPUT_PATH")));

        int stringsCount = Integer.parseInt(bufferedReader.readLine().trim());

        List<String> strings = new ArrayList<>();

        for (int i = 0; i < stringsCount; i++) {
            String stringsItem = bufferedReader.readLine();
            strings.add(stringsItem);
        }

        int queriesCount = Integer.parseInt(bufferedReader.readLine().trim());

        List<String> queries = new ArrayList<>();

        for (int i = 0; i < queriesCount; i++) {
            String queriesItem = bufferedReader.readLine();
            queries.add(queriesItem);
        }

        List<Integer> res = Result.matchingStrings(strings, queries);

        for (int i = 0; i < res.size(); i++) {
            bufferedWriter.write(String.valueOf(res.get(i)));

            if (i != res.size() - 1) {
                bufferedWriter.write("\n");
            }
        }

        bufferedWriter.newLine();

        bufferedReader.close();
        bufferedWriter.close();
    }
}
